src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -810,53 +899,66 @@ CopyLocalDataIntoShards(Oid distributedRelationId)
 	econtext->ecxt_scantuple = slot;
 
 	copyDest =
-		(DestReceiver *) CreateCitusCopyDestReceiver(distributedRelationId,
+		(DestReceiver *) CreateCitusCopyDestReceiver(destinationDistributedRelationId,
 													 columnNameList, estate,
 													 stopOnFailure);
 
 	/* initialise state for writing to shards, we'll open connections on demand */
 	copyDest->rStartup(copyDest, 0, tupleDescriptor);
 
-	/* begin reading from local table */
-	scan = heap_beginscan(distributedRelation, GetActiveSnapshot(), 0, NULL);
-
-	oldContext = MemoryContextSwitchTo(GetPerTupleMemoryContext(estate));
-
-	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	/* iterate on the local relations, copy each into the destination relation */
+	foreach(sourceRelationCell, sourceRelationList)
```

Comment:
`CreateHashDistributedTable` calls `ExecuteRemoteQuery` to create shards, which calls `MultiClientConnectStart` to open a connection. `MultiClientConnectStart` is not allowed after a modification (e.g. CopyLocalDataIntoShards) since it might not use the right placement connection.

However, you should be able to call `CreateHashDistributedTable` for all tables, and then `CopyLocalDataIntoShards` for all tables.